io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -519,10 +523,23 @@ newtype ClockId   = ClockId   Int deriving (Eq, Ord, Enum, Show)
 
 type ThreadLabel = String
 
+data LabeledThread = LabeledThread {
+    labeledThreadId    :: ThreadId,
+    labeledThreadLabel :: Maybe ThreadLabel
+  }
+  deriving (Eq, Ord, Show)
+
+labeledThreads :: Map ThreadId (Thread s a) -> [LabeledThread]
+labeledThreads =
+    Map.foldr
+      (\Thread { threadId, threadLabel } acc -> LabeledThread threadId threadLabel : acc)
+      []
```

Comment:
Why an explicit `foldr` instead of `Map.elems` + a list comprehension or a `map`?

Reply:
no big reason here; also `foldr` is as clear as `Map.elems` and list comprehension in this case.